libraries/AP_Mission/AP_Mission_ChangeDetector_Copter.cpp

```diff
@@ -0,0 +1,168 @@
+    const bool cmd1_is_wp = (mis_change_detect.cmd[1].id == MAV_CMD_NAV_WAYPOINT ||
+                             mis_change_detect.cmd[1].id == MAV_CMD_NAV_LOITER_UNLIM ||
+                             mis_change_detect.cmd[1].id == MAV_CMD_NAV_LOITER_TIME);
+    const bool cmd1_is_spline = (mis_change_detect.cmd[1].id == MAV_CMD_NAV_SPLINE_WAYPOINT);
+    const bool cmd1_is_other = (!cmd1_is_wp && !cmd1_is_spline);
+    const bool cmd1_existed = cmd_list_bak.cmd_count >= 2;
+    const bool cmd1_was_wp = (cmd_list_bak.cmd[1].id == MAV_CMD_NAV_WAYPOINT ||
+                             cmd_list_bak.cmd[1].id == MAV_CMD_NAV_LOITER_UNLIM ||
+                             cmd_list_bak.cmd[1].id == MAV_CMD_NAV_LOITER_TIME);
+    const bool cmd1_was_spline = (cmd_list_bak.cmd[1].id == MAV_CMD_NAV_SPLINE_WAYPOINT);
+    const bool cmd1_was_other = (!cmd1_was_wp && !cmd1_was_spline);
+
+    // if 2nd segment forces a pause then return NONE
+    if ((!cmd1_existed || cmd1_was_other) && (!cmd1_exists || cmd1_is_other)) {
+        debug("check_for_mission_change: 1st is forced pause, None");
+        return AP_Mission_ChangeDetector_Copter::ChangeResponseType::NONE;
+    }
+
+    // check for 1st is wp (without a pause), if add of 2nd wp then ADD_NEXT_WAYPOINT
+    const bool cmd1_added = (cmd_list_bak.cmd_count == 1) && (mis_change_detect.cmd_count > 1);
+
+    if (cmd0_is_wp && cmd1_added) {
+        if (cmd1_is_wp) {
+            // 2nd is wp
+            debug("check_for_mission_change: 1st is wp, no pause, 2nd added, AddNextWP");
+            return AP_Mission_ChangeDetector_Copter::ChangeResponseType::ADD_NEXT_WAYPOINT;
+        } else if (cmd1_is_spline) {
+            // 2nd is spline
+            // Currently we don't support set_destination_speed_max after leg has been started
+            // Therefore we can't add a spline without reseting
+            debug("check_for_mission_change: 1st is wp, no pause, 2nd added but is spline, Reset");
+            return AP_Mission_ChangeDetector_Copter::ChangeResponseType::RESET_REQUIRED;
+        } else {
+            // 2nd is nether wp or spline
+            // should be handled by pause above so we should not reach here
+            debug("check_for_mission_change: 1st is wp, no pause, 2nd added but not wp or spline, None");
+            return AP_Mission_ChangeDetector_Copter::ChangeResponseType::NONE;
+        }
+    }
+
+    // trajectory may be dependent on multiple waypoints, handle changed waypoints
+    if (cmd0_is_wp) {
+        // 1st is wp
+        if (!cmd1_exists) {
+            // 2nd deleted, Reset
+            // This could be extended to check cmd_list_bak for the deleted waypoint and return REMOVE_NEXT_WAYPOINT
+            debug("check_for_mission_change: 1st is wp, deleted 2nd wp, Reset");
+            return AP_Mission_ChangeDetector_Copter::ChangeResponseType::RESET_REQUIRED;
+        } else if (cmd1_is_wp && cmd1_was_wp) {
```

Comment:
This would be less intimidating if we didn't use `else in all of these cases.  It's causing a lot of unnecessary indenting.